cmd/sign/main.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2022 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"context"
+	"crypto"
+	"encoding/base64"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"syscall"
+
+	sigstore "github.com/sigstore/sigstore/pkg/signature"
+	"github.com/sigstore/sigstore/pkg/signature/kms"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"github.com/tektoncd/pipeline/pkg/reconciler/trustedresources"
+	"golang.org/x/term"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/webhook/resourcesemantics"
+	"sigs.k8s.io/yaml"
+)
+
+var (
+	keyFile      = flag.String("kf", "", "private key file path")
+	kmsKey       = flag.String("kms", "", "kms key path")
+	resourceFile = flag.String("rf", "", "YAML file path for tekton resources")
+	kind         = flag.String("kd", "Task", "The kind of the signed object. Supported values: [Task, Pipeline]")
+	targetDir    = flag.String("td", "", "Dir to save the signed files")
+	targetFile   = flag.String("tf", "signed.yaml", "Filename of the signed file")
+	// Read is for fuzzing
+	read = readPasswordFn
+)
+
+func main() {
+	ctx := context.Background()
+
+	flag.Parse()
+
+	var signer sigstore.Signer
+	var err error
+	if *keyFile != "" {
+		// Load signer from key files
+		signer, err = sigstore.LoadSignerFromPEMFile(*keyFile, crypto.SHA256, getPass)
+		if err != nil {
+			log.Fatalf("error getting signer: %v", err)
+		}
+	}
+	if *kmsKey != "" {
+		signer, err = kms.Get(ctx, *kmsKey, crypto.SHA256)
```

Comment:
s/s KMS doesn't work quite the same way as cosign since it's trying to minimize deps - you need to import all of the desired providers from https://github.com/sigstore/sigstore/tree/main/pkg/signature/kms manually, which will call [ProviderInit](https://pkg.go.dev/github.com/sigstore/sigstore/pkg/signature/kms#ProviderInit) to register the prefix.

The reason for this is that the providers are a big source of the massive dependency creep in cosign and related tools, since the cloud providers will pull in their corresponding SDKs.

I don't think we want to put this dependency burden on pipelines - at the very least you should make this its own module, though you may find it easier to just break this out into its own repo in the tektoncd org.

Reply:
Followed up off-thread - KMS dependency is likely unavoidable since we'll need it on the reconciler side anyway.

Import away 🙈

Reply:
Would still recommend a separate repo for this long term though - you'll likely find it easier to maintain / distribute.

Reply:
Maybe I remove those deps from this PR? I just tried to push them in a new commit and added 13k lines of code. 😢 
Then this commit doesn't really work for kms. (Can be added later when we have a separate repo)

Reply:
Yeah that sounds about right. It's kinda unavoidable given that you're pulling in the SDKs.